False positive on typed constant error

Hello,

First of all, thanks for open sourcing this! I'm testing it out in some codebases and it seems to be very capable of detecting lots of error already.

I have a codebase where using type wrappers on top of strings seems to trigger a false positive. Minimal reproducing program:

```go
package main

type Error string

func (err Error) Error() string { return string(err) }

const AnError Error = "this is an error"

func action(x bool) (*bool, error) {
	if x {
		return nil, AnError
	}
	var b bool
	return &b, nil
}

func useAction() error {
	ptr, err := action(true)
	if err != nil {
		return err
	}
	_ = *ptr
	return nil
}
```

Results in this analysis:
```sh
❯ nilaway .
/tmp/nilawayt/t.go:22:7: error: Potential nil panic detected. Observed nil flow from source to dereferenc
e point: 
        -> nilawayt/t.go:11:10: literal `nil` returned from `action()` in position 0 when the error retur
n in position 1 is not guaranteed to be non-nil through all paths
        -> nilawayt/t.go:22:7: result 0 of `action()` dereferenced

```

--- a/assertion/function/assertiontree/backprop_util.go
+++ b/assertion/function/assertiontree/backprop_util.go
@@ -245,9 +245,18 @@ func isErrorReturnNil(rootNode *RootAssertionNode, errRet ast.Expr) bool {
 
 // isErrorReturnNonnil returns true if the error return is guaranteed to be nonnil, false otherwise
 func isErrorReturnNonnil(rootNode *RootAssertionNode, errRet ast.Expr) bool {
+	// The error value's type is not inhabited by nil (e.g., a named basic type such as
+	// `type Error string`), so it can never be nil.
+	if rootNode.Pass().ExprBarsNilness(errRet) {
+		return true
+	}
+	// The error value is a custom error type (struct or pointer-to-struct), e.g. a returned
+	// `&MyErr{}` or `MyErr{}`. Such values are assumed non-nil: a concrete value boxed into the
+	// error interface yields a non-nil interface.
 	if t := rootNode.Pass().TypesInfo.TypeOf(errRet); typeshelper.AsDeeplyStruct(t) != nil {
 		return true
 	}
+	// The error value is the return of a hook-modeled function known to produce a non-nil error.
 	if callExpr, ok := errRet.(*ast.CallExpr); ok {
 		if hook.AssumeReturn(rootNode.Pass(), callExpr) != nil {
 			return true

--- a/testdata/src/go.uber.org/errorreturn/errorreturn.go
+++ b/testdata/src/go.uber.org/errorreturn/errorreturn.go
@@ -838,3 +838,53 @@ func Wrapf(e error) error {
 	}
 	return fmt.Errorf("wrapped: %w", e)
 }
+
+// ***** the below test cases check that an error value whose type cannot be inhabited by nil (e.g.,
+// a named basic type such as `type Error string`) is correctly treated as a non-nil error. This
+// addresses the false positive reported in https://github.com/uber-go/nilaway/issues/108 *****
+
+type typedConstErr string
+
+func (e typedConstErr) Error() string { return string(e) }
+
+const anError typedConstErr = "this is an error"
+
+// retTypedConstErr returns a typed-constant error in the non-nil error position. Since
+// `typedConstErr` is a named basic type that nil cannot inhabit, `anError` is a valid non-nil error
+// that suppresses consumption of the (nil) non-error result, so no error is expected here.
+func retTypedConstErr(x bool) (*int, error) {
+	if x {
+		return nil, anError
+	}
+	i := 0
+	return &i, nil
+}
+
+// callRetTypedConstErr exercises the exact pattern from issue #108: the result is dereferenced only
+// on the non-error path, which must not be flagged as a potential nil panic.
+func callRetTypedConstErr() error {
+	ptr, err := retTypedConstErr(true)
+	if err != nil {
+		return err
+	}
+	print(*ptr)
+	return nil
+}
+
+// retTypedErrVar checks the same behavior when the error flows through a variable of the named basic
+// type rather than being returned directly.
+func retTypedErrVar(x bool) (*int, error) {
+	var e typedConstErr = "boom"
+	if x {
+		return nil, e
+	}
+	i := 0
+	return &i, nil
+}
+
+func callRetTypedErrVar() {
+	ptr, err := retTypedErrVar(true)
+	if err == nil {
+		print(*ptr)
+	}
+}

